### PR TITLE
feat: encoder structure

### PR DESCRIPTION
**Pull request by davidbp**

### CODE REVIEW

File: vggish_audio_encoder.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        self.sess = tf.compat.v1.Session()
+        define_vggish_slim()
+        load_vggish_slim_checkpoint(self.sess, self.model_path)
+        self.feature_tensor = self.sess.graph.get_tensor_by_name(
+            INPUT_TENSOR_NAME)
+        self.embedding_tensor = self.sess.graph.get_tensor_by_name(
+            OUTPUT_TENSOR_NAME)
+        self.post_processor = Postprocessor(self.pca_path)
+
+    def post_init(self):
```
Comment (bwanglzu) [inline vggish_audio_encoder.py]: dont forget to remove `post_init` and merge into `init` if necessary

**Review by bwanglzu — CHANGES_REQUESTED** [review]

File: vggish_audio_encoder.py (hunk 1 above)
Comment (davidbp) [reply vggish_audio_encoder.py]: yes, removed.

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,9 @@
-torch==1.7.1
-clip @ git+https://github.com/openai/CLIP.git@cfcffb90e69f37bf2ff1e988237a0fbe41f33c04
 jina==2.0.0rc6
+resampy==0.2.2
+tensorflow_cpu==2.2.0
+tf_slim==1.1.0
+six==1.15.0
+soundfile==0.10.3.post1
+librosa==0.8.0
+click==7.1.2
```
Comment (bwanglzu) [inline requirements.txt]: why click needed?

File: scripts/download_model.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+DATA_DIR=data
+REPO_DIR=audioset_tagging_cnn
+
+# === Check for dirs first and get errors out of the way ===
+if [ -d "$MODEL_DIR" ]; then
+  echo $MODEL_DIR ' exists. Please delete it to continue'
+  exit 1
+fi
+
+# ============ Download the model ============
```
Comment (bwanglzu) [inline scripts/download_model.sh]: remove the comments because they're in `echo`

File: tests/integration/test_exec.py [hunk 4]
```diff
@@ -1,12 +1,35 @@
...
 
-def test_exec():
-    f = Flow().add(uses=MyDummyExecutor)
-    with f:
-        resp = f.post(on='/test', inputs=Document(), return_results=True)
-        assert resp is not None
+try:
+    from vggish_audio_encoder import VggishAudioEncoder
+except:
+    from jinahub.encoders.audio.vggish_audio_encoder import VggishAudioEncoder
```
Comment (bwanglzu) [inline tests/integration/test_exec.py:13]:
```suggestion
from jinahub.encoders.audio.vggish_audio_encoder import VggishAudioEncoder
```

File: tests/integration/test_exec.py [hunk 5]
```diff
@@ -1,12 +1,35 @@
...
+
+
+def test_flow_from_yml():
+
+    doc = DocumentArray([Document()])
+    with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+        resp = f.post(on='test', inputs=doc, return_results=True)
+
+    assert resp is not None
+
```
Comment (bwanglzu) [inline tests/integration/test_exec.py:25]: more tests such as size of the output embedding?

**Review by bwanglzu — CHANGES_REQUESTED** [review]

File: tests/integration/test_exec.py [hunk 6]
```diff
@@ -1,12 +1,35 @@
...
+def test_embedding():
+    x_audio, sample_rate = librosa.load(os.path.join(cur_dir, '../data/sample.wav'))
+    log_mel_examples = vggish_input.waveform_to_examples(x_audio, sample_rate)
+    doc = DocumentArray([Document(blob=log_mel_examples)])
+
+    with Flow.load_config(os.path.join(cur_dir, 'flow.yml')) as f:
+        responses = f.post(on='index', inputs=doc, return_results=True)
+
+    assert responses[0].docs[0].embedding is not None
```
Comment (Yongxuanzhang) [inline tests/integration/test_exec.py:36]: maybe assert the shape as well?

File: tests/unit/test_exec.py [hunk 7]
```diff
@@ -1,9 +1,17 @@
...
+
+try:
+    from vggish_audio_encoder import VggishAudioEncoder
+except:
+    from jinahub.encoders.audio.vggish_audio_encoder import VggishAudioEncoder
+
+cur_dir = os.path.dirname(os.path.abspath(__file__))
+
+def test_load():
+    encoder = Executor.load_config(os.path.join(cur_dir, '../../config.yml'))
```
Comment (Yongxuanzhang) [inline tests/unit/test_exec.py:17]: do we need more unit tests?

File: requirements.txt (hunk 2 above)
Comment (JoanFM) [reply requirements.txt]: right, click should not be needed

Comment (JoanFM) [reply requirements.txt]: or pandas either. If this is needed for testing add it in the local requirements file of the test folder

File: tests/integration/test_exec.py (hunk 4 above)
Comment (JoanFM) [reply tests/integration/test_exec.py:13]: @bwanglzu if u do this, do the tests pass locally for u?

File: tests/unit/test_exec.py (hunk 7 above)
Comment (JoanFM) [reply tests/unit/test_exec.py:17]: Yes, most of the tests should be unit tests

File: vggish_audio_encoder.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+                 default_traversal_paths: Iterable[str] = ['r'],
+                 *args, **kwargs):
+
+        super().__init__(*args, **kwargs)
+        self.model_path = model_path
+        self.pca_path = pca_path
+        self.default_batch_size = default_batch_size
+        self.default_traversal_paths = default_traversal_paths
+
+        self.sess = tf.compat.v1.Session()
```
Comment (JoanFM) [inline vggish_audio_encoder.py:37]: I think this session must be closed at exit I would say?

File: vggish_audio_encoder.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+        return _batch_generator(filtered_docs, batch_size)
+
+    def _create_embeddings(self, document_batches_generator: Iterable):
+        """Update the documents with the embeddings generated by VGGISH"""
+
+        for document_batch in document_batches_generator:
+            for d in document_batch:
+                # Vggish broadcasts across different length audios, not batches
+                [embedding] = self.sess.run([self.embedding_tensor], feed_dict={self.feature_tensor: d.blob})
```
Comment (JoanFM) [inline vggish_audio_encoder.py]: Does not make much sense to have `batch_size` if then the embedding is computed one by one.

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: scripts/download_model.sh (hunk 3 above)
Comment (davidbp) [reply scripts/download_model.sh]: done

File: tests/integration/test_exec.py (hunk 4 above)
Comment (davidbp) [reply tests/integration/test_exec.py:13]: They do not pass locally if I do this

File: vggish_audio_encoder.py (hunk 9 above)
Comment (davidbp) [reply vggish_audio_encoder.py]: good point, updated

File: vggish_audio_encoder.py (hunk 8 above)
Comment (davidbp) [reply vggish_audio_encoder.py:37]: Yes, updated

File: tests/unit/test_exec.py (hunk 7 above)
Comment (davidbp) [reply tests/unit/test_exec.py:17]: added the dimension test that Yongxuanzhang mentioned

File: tests/integration/test_exec.py (hunk 6 above)
Comment (davidbp) [reply tests/integration/test_exec.py:36]: done as unit test

File: tests/integration/test_exec.py (hunk 5 above)
Comment (davidbp) [reply tests/integration/test_exec.py:25]: Done as unit test

File: tests/integration/flow.yml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+jtype: Flow
+version: '1'
```
Comment (JoanFM) [inline tests/integration/flow.yml:2]: this integration flow will need to add the `py_modules`

File: requirements.txt [hunk 11]
```diff
@@ -1,3 +1,7 @@
-torch==1.7.1
-clip @ git+https://github.com/openai/CLIP.git@cfcffb90e69f37bf2ff1e988237a0fbe41f33c04
 jina==2.0.0rc6
```
Comment (JoanFM) [inline requirements.txt]: try with 2.0.0.rc10

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 11 above)
Comment (davidbp) [reply requirements.txt]: updated

File: tests/integration/flow.yml (hunk 10 above)
Comment (davidbp) [reply tests/integration/flow.yml:2]: updated

**Pull request merged**